xnmt/translator.py

```diff
@@ -277,9 +292,23 @@ def __call__(self, translator, dec_state, src, trg):
       except ValueError:
         pass
       # Calculate the evaluation score
-      eval_score.append(self.evaluation_metric.evaluate_fast(trg_i.words, sample_i))
-
-    return -dy.sum_elems(dy.cmult(dy.inputTensor(eval_score, batched=True), dy.esum(logsofts)))
+      score = 0 if not len(sample_i) else self.evaluation_metric.evaluate_fast(trg_i.words, sample_i)
+      self.eval_score.append(score)
+      self.trueScore = dy.inputTensor(self.eval_score, batched=True)
+
+    final_score = -self.trueScore
+    if self.use_baseline:
+      final_score = self.bs - self.trueScore
+    return dy.sum_elems(dy.cmult(final_score , dy.esum(logsofts)))
+
+  @recursive_sum
+  def calc_additional_loss(self, _):
```

Comment:
I don't think you need to do this. This is required if you are calculating the reward. In this case, the loss can be composed inside the __call__ and the LossBuilder() object is returned instead.

Reply:
Oh, I am confused here, @philip30. Are you suggesting to make the standard-loss an instance of LossBuilder? or am I misunderstanding something? Currently, I guess xnmt_train the "standard loss" (from the model's calc_loss) is considered by default to be a single loss and all additional losses seem to go as an instance of LossBuilder.  Wouldn't further generalizing this format make things more confusing?

Reply:
You can do this to substitute the return in __call__:
``` python
loss = LossBuilder()
loss.add_loss("reinf_loss", dy.sum_elems(dy.cmult(final_score , dy.esum(logsofts))))
loss.add_loss("baseline", dy.sum_batches(dy.squared_distance(self.trueScore, self.bs))
return loss
```
It doesn't matter if you return a loss_builder or an instance of dynet_expression as far as I know.

Reply:
Hi thanks a lot for you quick replies.
When I do this, I get the following error:
```     
   File "/xnmt/xnmt/xnmt_train.py", line 231, in run_epoch
   additional_loss = self.model.calc_additional_loss(dy.nobackprop(-standard_loss))   
   TypeError   :    bad operand type for unary -: 'LossBuilder'  
```

So, it seems that main algorithm doesn't expect LossBuilder object there. I can change it. But I wanted to ask if it is a good design choice.

Reply:
I see. This is a new case, but I think the calc_loss should handle both the LossBuilder and dynet expression object. You can fix it with a design you think is good.
I think you can either implement the unary operator in LossBuilder or add a conditional statment there.

Reply:
Or if it is too difficult, you can neglect the LossBuilder. Instead just use the "+" to sum 2 dynet expression. LossBuilder is used only for reporting purposes (so we can see 2 losses are going down separately)